### DISCUSSION TITLE
Prevent `handleSubmit` from reseting `formState.errors`?

### DISCUSSION THREAD

**Discussion by bestickley** [react-hook-form/react-hook-form#9904] [Q&A] [answered]
When I submit my form and `handleSubmit` calls my `onSubmit` function, sometimes I short circuit the submission function and return early. The issue is when I return early in the `onSubmit` function, the errors in `formState.errors` reset. I don't want them to reset. Is this expected behavior?

My current work around is do something like this which is silly:
```ts
const onSubmit = useCallback(
    (data: ProductBillingSchema) => {
      if (formState.errors.costObjectId) {
        // must set error otherwise previous error will be reset
        setError("costObjectId", formState.errors.costObjectId);
        return;
      }
     ...
)]
```

**Comment by leapful** [comment, marked as answer]
@bestickley 

Could you please provide a codesandbox to re-produce the issue ? 

If you are using `handleSubmit` method from `useForm` hook, the callback `onValid` from `handleSubmit(onValid)` is only triggered if there is no error.

**Reply by bestickley** [reply]
@leapful, thank you for the response. I tried setting up a CSB, but ran into a weird issue. Please see [here](https://codesandbox.io/s/kind-cherry-dmovxs?file=/src/App.tsx). Try clicking the "Set Error" button and you can see in logs it's set but then reset by something? I don't understand why. What I wanted to show you is that if you click "Set Error" and the error appears, and then you click submit but return early in `onSubmit`, the error disappears but I want it to be preserved. But I wasn't able to show that because of the previously mentioned issue.

I think I've found the issue. When `handleSubmit` is called if a resolver is specified, then the manually set errors are discarded with the schema errors. See [this line](https://github.com/react-hook-form/react-hook-form/blob/master/src/logic/createFormControl.ts#L1038) where the code is `_formState.errors = errors;`. Shouldn't it be `_formState.errors = { ...errors, .._formState.errors }` so that manually set errors are preserved?

**Reply by leapful** [reply]
@bestickley 

The button "Set Error" has no attribute "type" therefore it will be use default "type=submit" and triggers submit event.

Just add `type="button"` to prevent submit the form on setting error via method `setError`.

https://codesandbox.io/s/silly-fast-3yheq9

**Reply by bestickley** [reply]
@leapful, thank you for fixing/explaining. Do you see the issue now? The error incorrectly resets when form is submitted. Is there something else I'm missing?

**Reply by leapful** [reply]
@bestickley 

There is no issue on your codesandbox example. You are validating field name `firstName` with zod as a string but by default, zod only checks mandatory condition, meaning not null or not undefined, while your default value of field `firstName` is an empty string. When submitting form, the validation is re-executed and empty string is a valid value following mandatory rule, therefore, it will clear error on field `firstName`.

Like I said on the previous comment, `handleSubmit(onValid)` has `onValid` is a callback that only triggers after re-validation and all form values are valid. You cannot add a check if else on form errors to stop the flow at `onValid` callback.

If you want to keep a custom error , please follow documentation https://react-hook-form.com/api/useform/seterror on setting custom error.

If you want to set custom error on a registered field like `firstName`, enhance `<form onSubmit />` before passing `handleSubmit(onValid)`, just remember that now you will need to set / clear error on the field manually by yourself. Something like:

```
const onSubmit = (event) => {
    event.preventDefault();
    event.stopPropagation();

   if (error) {
       // do something
       return
   }

   handleSubmit(onValid)(event);
}
```

Or, you can apply the most simple solution is to update zod validation rule on field `firstName` to not allow empty string, like this example https://codesandbox.io/s/silly-fast-3yheq9?file=/src/App.tsx

**Reply by bestickley** [reply]
> If you want to keep a custom error , please follow documentation https://react-hook-form.com/api/useform/seterror on setting custom error.

This is the key I was looking for. The custom function above where you call `handleSubmit` manually is also a great idea. Thank you, @leapful! Really appreciate your patience.
